velox/functions/prestosql/tests/MapFromEntriesTest.cpp

```diff
@@ -0,0 +1,161 @@
+}
+
+TEST_F(MapFromEntriesTest, nullMapEntries) {
+  auto rowType = ROW({INTEGER(), INTEGER()});
+  std::vector<std::vector<variant>> data = {
+      {variant(TypeKind::ROW)}, {variant::row({1, 11})}};
+  auto input = makeArrayOfRowVector(rowType, data);
+  VELOX_ASSERT_THROW(
+      evaluateExpr("map_from_entries(C0)", {input}),
+      "map entry cannot be null");
+  auto expected =
+      makeNullableMapVector<int32_t, int32_t>({std::nullopt, O({{1, 11}})});
+  verifyMapFromEntriesWithTry(input, expected);
+}
+
+TEST_F(MapFromEntriesTest, nullKeys) {
+  auto rowType = ROW({INTEGER(), INTEGER()});
+  std::vector<std::vector<variant>> data = {
+      {variant::row({variant::null(TypeKind::INTEGER), 0})},
+      {variant::row({1, 11})}};
+  auto input = makeArrayOfRowVector(rowType, data);
+  VELOX_ASSERT_THROW(
+      evaluateExpr("map_from_entries(C0)", {input}), "map key cannot be null");
+  auto expected =
+      makeNullableMapVector<int32_t, int32_t>({std::nullopt, O({{1, 11}})});
+  verifyMapFromEntriesWithTry(input, expected);
+}
+
+TEST_F(MapFromEntriesTest, duplicateKeys) {
+  auto rowType = ROW({INTEGER(), INTEGER()});
+  std::vector<std::vector<variant>> data = {
+      {variant::row({1, 10}), variant::row({1, 11})}, {variant::row({2, 22})}};
+  auto input = makeArrayOfRowVector(rowType, data);
+  VELOX_ASSERT_THROW(
+      evaluateExpr("map_from_entries(C0)", {input}),
+      "Duplicate map keys (1) are not allowed");
+  auto expected =
+      makeNullableMapVector<int32_t, int32_t>({std::nullopt, O({{2, 22}})});
+  verifyMapFromEntriesWithTry(input, expected);
+}
+
+TEST_F(MapFromEntriesTest, nullValues) {
+  auto rowType = ROW({INTEGER(), INTEGER()});
+  std::vector<std::vector<variant>> data = {
+      {variant::row({1, variant::null(TypeKind::INTEGER)}),
+       variant::row({2, 22}),
+       variant::row({3, 33})}};
+  auto input = makeArrayOfRowVector(rowType, data);
+  auto expected =
+      makeMapVector<int32_t, int32_t>({{{1, std::nullopt}, {2, 22}, {3, 33}}});
+  verifyMapFromEntries(input, expected);
+}
+
+TEST_F(MapFromEntriesTest, constantArray) {
+  auto rowType = ROW({VARCHAR(), INTEGER()});
+  std::vector<std::vector<variant>> data = {
+      {variant::row({"red", 1}),
+       variant::row({"blue", 2}),
+       variant::row({"green", 3})},
+      {variant::row({"red shiny car ahead", 4}),
+       variant::row({"blue clear sky above", 5})},
+      {variant::row({"r", 11}),
+       variant::row({"g", 22}),
+       variant::row({"b", 33})}};
+  auto input = makeArrayOfRowVector(rowType, data);
+  auto expected = {
+      makeMapVector<StringView, int32_t>(
+          {{{"red"_sv, 1}, {"blue"_sv, 2}, {"green"_sv, 3}}}),
+      makeMapVector<StringView, int32_t>(
+          {{{"red shiny car ahead"_sv, 4}, {"blue clear sky above"_sv, 5}}}),
+      makeMapVector<StringView, int32_t>(
+          {{{"r"_sv, 11}, {"g"_sv, 22}, {"b"_sv, 33}}})};
+
+  evaluateExprWithConstantArray(expected, "map_from_entries(C0)", input);
+}
```

Comment:
Could you add a unit test where the row vector inside the array vector and the keys vector are not flat vectors?

Reply:
Hi @darrenfu, It's different from evaluateExprWithConstantArray() which only make the top-level array vector in constant encoding. What I was suggesting is that since applyFlat handles different encoding of the element vector (i.e., the row vector) of the input array vector, it'd be better to have a test for this too.

Reply:
@kagamiori - I've added two new unit test cases to cover your case (`dictionaryEncodedElementsInFlat`) and the case  (`outputSizeIsBoundBySelectedRows`) where I fixed a bug (similar to #4455 ).